File: 백준/Gold/1477. 휴게소 세우기/휴게소 세우기.js
const readline = require('readline');
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});
let lineCount = 0;
let n, m, l;
let arr = [];
rl.on('line', (line) => {
    if (lineCount === 0) {
        [n, m, l] = line.split(' ').map(Number);
        arr = [0, l];
        if (n === 0) {
            solveProblem();
        }
    } else if (lineCount === 1) {
        if (n > 0) {
            arr = arr.concat(line.split(' ').map(Number));
            solveProblem();
        }
    }
    lineCount++;
});

function solveProblem() {
    arr.sort((a, b) => a - b);
    
    let start = 0;
    let end = l;
    
    while (start + 1 < end) {
        const mid = (start + end) >> 1;
        let count = 0;
        
        for (let i = 0; i < arr.length - 1; i++) {
            const dist = arr[i + 1] - arr[i];
            count += Math.ceil(dist / mid) - 1;
        }
        
        if (count > m) {
            start = mid;
        } else {
            end = mid;
        }
    }
    
    console.log(end);
    rl.close();
}

File: 백준/Gold/1477. 휴게소 세우기/휴게소 세우기.java
import java.io.*;
import java.util.*;

public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;
        st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        int l = Integer.parseInt(st.nextToken());

        st = new StringTokenizer(br.readLine());
        List<Integer> arr = new ArrayList<>();
        arr.add(0);
        arr.add(l);
        while (st.hasMoreTokens()) arr.add(Integer.parseInt(st.nextToken()));
        Collections.sort(arr);
        int start = 1;
        int end = l;
        while (start <= end) {
            int mid = (start + end) / 2;
            int count = 0;
            for (int i = 0; i <= n; i++) {
                int dist = arr.get(i+1) - arr.get(i) - 1;
                count += dist / mid;
            }
            if (count > m) start = mid + 1;
            else end = mid - 1;
        }
        System.out.println(start);
    }

}